martian-derive/tests/mro/test_typed_map.mro

```diff
@@ -0,0 +1,20 @@
+#
+# Copyright (c) 2020 10X Genomics, Inc. All rights reserved.
+#
+# WARNING: This file is auto-generated.
+# DO NOT MODIFY THIS FILE DIRECTLY
+#
+
+filetype fastq;
+
+struct ReadsStruct(
+    string     name,
+    map<fastq> reads_map,
+    map[]      multi_reads_map,
```

Comment:
This isn't correct.  It should be
```suggestion
    map<fastq>[]      multi_reads_map,
```
You should also add a test case for `map<fastq[]>` and maybe some multi-dimensional array types.

Reply:
Yeah I haven't yet figured out a good way to handle arrays of typed maps. Currently it prints them as untyped. The reason is that it sort of breaks the hierarchy of types MartianRust assumes: 
```
pub enum MartianBlanketType {
    Primary(MartianPrimaryType),
    Array(MartianPrimaryType),
    TypedMap(MartianPrimaryType),
}
```
An untyped map is a primary type so you can have an array of them, but you can't have an array of a TypedMap since that's a blanket type. Perhaps the best solution would be to make ArrayOfTypedMap(MartianPrimaryType) its own blanket type here.

Martian-Rust doesn't support multidimensional arrays at the moment, although it would probably be easy to similarly add a TwoDimensionalArray(MartianPrimaryType) which serializes to and from `Vec<Vec<T>>`. `map<fastq[]>` would similarly need its own blanket type probably. @nlhepler I can't think of any other solutions, does this seem reasonable to you?

Reply:
I'm not familiar enough with rust's type system but is type recursion not allowed, e.g. can you not have `Array(MartianBlanketType)`?

Reply:
A quick google search suggests you could have e.g. `Array(Box<MartianBlanketType>)`.

Reply:
Regarding the trait bounds on the key, you're totally correct -- I was able to remove those bounds. I think these bounds may have been required a few commits back when my less-clean solution involved wrapping the HashMap type.

Reply:
Ohh. I think you're right

Reply:
Worked like a charm! Awesome.

Reply:
Nice.  One thing I don't see in the test cases, however, is `map<StructType>`.  I assume that should work fine, since `StructType[]` works.

Reply:
Also, might want to double-check that the `Box` is actually necessary.  If it works on its own or with `Option<T>` that might be better; the examples I've seen where `Box<T>` was used were cases where an enum contains a direct recursive link, whereas an array or map might be fine.

Reply:
Unless I'm mistaken this is a direct recursive link. There's no actual array or map contained in this type, it's just a martian type definition, and basically removing the box makes it have infinite size since it wants to statically store a full MartianBlanketType inside, and each of those contains a MartainBlanketType, and there's the overhead of the struct itself so the size blows up.
```
error[E0072]: recursive type `mro::MartianBlanketType` has infinite size
   --> martian/src/mro.rs:183:1
    |
183 | pub enum MartianBlanketType {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ recursive type has infinite size
184 |     Primary(MartianPrimaryType),
185 |     Array(MartianBlanketType),
    |           ------------------ recursive without indirection
186 |     TypedMap(MartianBlanketType),
    |              ------------------ recursive without indirection
    |
help: insert some indirection (e.g., a `Box`, `Rc`, or `&`) to make `mro::MartianBlanketType` representable
    |
185 |     Array(Box<MartianBlanketType>),
186 |     TypedMap(Box<MartianBlanketType>),
    |

error[E0391]: cycle detected when computing drop-check constraints for `mro::MartianBlanketType`
   --> martian/src/mro.rs:183:1
    |
183 | pub enum MartianBlanketType {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: ...which again requires computing drop-check constraints for `mro::MartianBlanketType`, completing the cycle
    = note: cycle used when type-checking all item bodies
```

Reply:
Ah, yes.  It would be possible to contain `Vec<MartianBlanketType>` (since a vector is actually [size plus] a pointer to an array) but I guess that's not actually what you want in this case.